### DISCUSSION TITLE
Edit in-memory certstore while proxy is running

### DISCUSSION THREAD

**Discussion by Ilke-dev** [mitmproxy/mitmproxy#5859] [Q&A] [answered]
Hi,

I have a question about the addonmanager and the tlsconfig addon. I'm trying to edit the in-memory certstore while the proxy is running. I try to achieve this registering the addon into my own custom addon as a attribute. Then i create the ```CertStoreEntry``` from an certificate chain in memory (PEM format key included).  

However it seems that the tlsconfig addon from the addonmanager is just a unused copy, when the proxy is running it is deepcopied and i cannot seem to find the actual certstore that is used through the addonmanager.

Can someone point me to the right direction how i can edit the in-memory certstore while the proxy is already is running?

Example code i'm using:

```python
# -*- coding: utf-8 -*-
from mitmproxy.tools.main import mitmweb
import uvloop

from mitmproxy.addonmanager import Loader
from mitmproxy.certs import CertStoreEntry, Cert
from cryptography.hazmat.primitives.serialization import load_pem_private_key


class CustomAddon:
    
    def load(self, loader: Loader):
        self.master = loader.master
        self.tlsconfig = loader.master.addons.lookup["tlsconfig"]
        self.tlsconfig.configure("confdir")

    async def running(self):
        cert = self.retrieve_cert()
        self.tlsconfig.certstore.add_cert(self.create_certstore_entry(cert))

    def create_certstore_entry(self, cert_chain):
        cert = Cert.from_pem(cert_chain.encode("utf-8"))
        key = load_pem_private_key(cert_chain.encode("utf-8"), None)
        return CertStoreEntry(cert, key, None)

    def retrieve_cert(self):
        # retrieving some cert from database as a PEM which includes the key at the end
        ...

uvloop.install()

addons = [CustomAddon()]

if __name__ == "__main__":
    # Initialize mitmweb
    mitmweb(args=[
                "--web-host",
                "0.0.0.0",
                "--mode",
                "regular",
                "--listen-host",
                "0.0.0.0",
                "--listen-port",
                "443",
                "-s",
                "customaddon.py", # point to the file it self
                "--ssl-insecure"
            ]
    )
```

**Comment by mhils** [comment, marked as answer]
> However it seems that the tlsconfig addon from the addonmanager is just a unused copy, when the proxy is running it is deepcopied and i cannot seem to find the actual certstore that is used through the addonmanager.

That shouldn't be the case really, the tlsconfig certstore is the correct one.

Briefly looking through the code, my main suspicion would be that this weird hack here creates a new certstore after you have added your cert:

https://github.com/mitmproxy/mitmproxy/blob/d06afcc1f98a04065c656197768c2b6f3ef40d0d/mitmproxy/addons/tlsconfig.py#L403-L406

Could you confirm if that's the case? Git blame says I authored that as part of a larger refactor, maybe we're luck and it's not true anymore and can go away.

**Reply by Ilke-dev** [reply]
Hmmm, so i always get a blank certstore whenever i add. So is this classified as a bug?

**Reply by Ilke-dev** [reply]
I just uncommented line 406 and retried, it works now.

**Reply by mhils** [reply]
If you could do the legwork of checking if that breaks any of the tools and passes the testsuite, I'd be more than happy to merge a PR! :)

**Reply by Ilke-dev** [reply]
It seems like it won't break anything
![image](https://user-images.githubusercontent.com/105493024/211205877-febdcbcc-ac05-44d7-b57c-fda7c8814d0b.png)

**Reply by Ilke-dev** [reply]
Is this sufficient or did you expect something else?
